File: samples/acquisition/capture-10-images/c/main.c

/* capture_images - XIMEA xiAPI capture sample (C)
 *
 * Opens the first available XIMEA camera, sets exposure to 100 ms,
 * captures 10 frames, prints per-frame metadata, then closes.
 *
 * Build: see CMakeLists.txt or build.ps1 at the repo root.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <xiApi.h>

const int frameCount = 10;
const int exposureUs = 100000;
const int grabTimeoutMs = 5000;

#define CE(func)                                       \
    do                                                 \
    {                                                  \
        XI_RETURN stat = (func);                       \
        if (stat != XI_OK)                             \
        {                                              \
            fprintf(stderr, "Error: %s returned %d\n", \
                    #func, (int)stat);                 \
            goto cleanup;                              \
        }                                              \
    } while (0)

int main(void)
{
    uint32_t count = 0;
    HANDLE cam = NULL;
    XI_IMG img;
    XI_RETURN st;
    int ret = EXIT_SUCCESS;

    CE(xiGetNumberDevices(&count));
    if (count == 0)
    {
        fprintf(stderr, "Error: no XIMEA cameras detected\n");
        return EXIT_FAILURE;
    }

    printf("Found %u camera(s), opening index 0\n", (unsigned)count);

    CE(xiOpenDevice(0, &cam));

    CE(xiSetParamInt(cam, XI_PRM_EXPOSURE, exposureUs));

    printf("Exposure: %d us (%d ms)\n", exposureUs, exposureUs / 1000);

    CE(xiStartAcquisition(cam));

    printf("Capturing %d frames\n", frameCount);

    for (int i = 0; i < frameCount; i++)
    {
        memset(&img, 0, sizeof(img));
        img.size = sizeof(img);

        CE(xiGetImage(cam, grabTimeoutMs, &img));

        printf("Frame %d/%d: %ux%u nframe=%u first_byte=%d\n", i + 1, frameCount,
               (unsigned)img.width, (unsigned)img.height, (unsigned)img.nframe,
               img.bp ? (int)((unsigned char *)img.bp)[0] : -1);
    }

cleanup:
    if (cam)
    {
        xiStopAcquisition(cam);
        xiCloseDevice(cam);
    }

    if (ret == EXIT_SUCCESS)
        printf("Done\n");

    return ret;
}

File: samples/acquisition/capture-10-images/python/main.py
"""main - XIMEA xiAPI capture sample (Python 3.9+)

Opens the first available XIMEA camera, sets exposure to 100 ms,
captures 10 frames, prints per-frame metadata, then closes.

Build: no build step needed — run directly with Python.
"""

import sys
from ximea import xiapi

frame_count = 10
exposure_us = 100000   # 100 ms in microseconds
grab_timeout_ms = 5000  # must exceed exposure


def run_capture(cam):
    """Set exposure, acquire frames, print per-frame metadata.

    Args:
        cam: Open xiapi.Camera instance.

    Returns:
        0 on success, 1 on failure.
    """
    cam.set_exposure(exposure_us)
    print(f"Exposure: {exposure_us} us ({exposure_us // 1000} ms)")

    cam.start_acquisition()
    print(f"Capturing {frame_count} frames")

    img = xiapi.Image()
    try:
        for i in range(frame_count):
            cam.get_image(img, timeout=grab_timeout_ms)
            data = img.get_image_data_numpy()
            first_byte = int(data.flat[0]) if data is not None and data.size > 0 else -1
            print(
                f"Frame {i + 1}/{frame_count}: "
                f"{img.width}x{img.height} "
                f"nframe={img.nframe} "
                f"first_byte={first_byte}"
            )
    except xiapi.Xi_error as e:
        print(f"Error: failed on frame {i + 1}/{frame_count}: {e}", file=sys.stderr)
        cam.stop_acquisition()
        return 1

    cam.stop_acquisition()
    print("Done")
    return 0


if __name__ == "__main__":
    cam = xiapi.Camera()

    count = cam.get_number_devices()
    if count == 0:
        print("Error: no XIMEA cameras detected", file=sys.stderr)
        sys.exit(1)

    print(f"Found {count} camera(s), opening index 0")

    try:
        cam.open_device()
    except xiapi.Xi_error as e:
        print(f"Error: could not open camera: {e}", file=sys.stderr)
        sys.exit(1)

    ret = run_capture(cam)
    cam.close_device()
    sys.exit(ret)
